lang/wasm_instance_cache.hh

```diff
@@ -0,0 +1,326 @@
+        if (!func) {
+            throw wasm::exception(format("Exported object {} is not a function", ctx.function_name));
+        }
+        auto memory_export = instance.get(store, "memory");
+        if (!memory_export) {
+            throw wasm::exception("memory export not found - please export `memory` in the wasm module");
+        }
+        auto memory = std::get<wasmtime::Memory>(*memory_export);
+
+        return wasm::wasm_instance{.store=std::move(store), .instance=std::move(instance), .func=std::move(*func), .memory=std::move(memory)};
+    }
+
+    // lru must not be empty, and its elements must refer to entries in _cache
+    void evict_lru() noexcept {
+        auto& entry = _lru.front();
+        _total_size -= entry.instance_size;
+        _lru.pop_front();
+        auto it = _cache.find(entry.key);
+        _cache.erase(it);
+    }
+
+    void on_timer() noexcept {
+        auto now = seastar::lowres_clock::now();
+        while (!_lru.empty() && _lru.front().timestamp + _timer_period < now) {
+            evict_lru();
+        }
+    }
+
+public:
+    seastar::future<value_type> get(const key_type& key, wasm::context& ctx) {
+        cache_key_type key_pair(seastar::current_scheduling_group(), key);
+        auto& i = _cache[key_pair];
+        if (!i.instance) {
+            ++shard_stats().cache_misses;
+            i.instance = seastar::make_lw_shared<wasm_instance>(load(ctx));
+            _lru.emplace_back(key_pair, seastar::lowres_clock::now(), i.instance->memory.size(i.instance->store));
+            i.it = --_lru.end();
+            _total_size += i.it->instance_size;
+        } else {
+            ++shard_stats().cache_hits;
+        }
+        // cache entry for key_pair shouldn't be removed until we 'recycle' the instance
+        if (!i.mutex.try_lock()) {
+            // the instance is in use
+            ++shard_stats().cache_blocks;
+        } else {
+            i.mutex.unlock();
+            // because we don't want to remove an instance in use, and also because we can't 
+            // track its size efficiently, we remove it from lru and subtract its size from
+            // the total size until it is no longer used
+            _total_size -= i.it->instance_size;
+            _lru.erase(i.it);
+        }
+        return i.mutex.lock().then([this, &i, &ctx] {
+            if (!i.instance) {
+                i.instance = seastar::make_lw_shared<wasm_instance>(load(ctx));
+            }
+            return make_ready_future<value_type>(i.instance);
+        });
+    }
+
+    void reset(const key_type& key) noexcept {
+        auto key_pair = lookup_cache_key_type(seastar::current_scheduling_group(), key);
+        auto it = _cache.find(key_pair);
+        it->second.instance = nullptr;
+    }
+
+    void recycle(const key_type& key) noexcept {
+        auto it = _cache.find(lookup_cache_key_type(seastar::current_scheduling_group(), key));
```

Comment:
It's better to give the caller an object rather than rely on a second lookup. We should have the same scheduling group, but with an object it's clearer and safer. We can also utilize its destructor to automatically do things.

Reply:
Yes. I think it's okay that we're intrusive here, the service and its user are deeply tied anyway.